### DISCUSSION TITLE
Bottlerocket support

### DISCUSSION THREAD

**Discussion by rcgeorge23** [nestybox/sysbox#487] [General] [answered]
Hi sysbox team,

I was wondering whether there are any plans to support [Bottlerocket](https://aws.amazon.com/bottlerocket/) in the future?

Thanks

**Comment by ctalledo** [comment, marked as answer]
Hi @rcgeorge23, thanks for opening this discussion. 

We've not had a chance yet at looking into running Sysbox on Bottlerrocket, but it would be a natural fit. The upcoming Sysbox v0.5.0 release will make it possible to run Sysbox on most distros assuming kernel >= 5.12 (i.e., to leverage a new feature called ID-mapped mounts that Sysbox uses to expose host files into the rootless container with proper permissions). Thus, if Bottlerocket brings in kernel >= 5.12, it's likely Sysbox v0.5.0 will work on it, although we would need to try and test (devil is in the details always :) ).

**Reply by rcgeorge23** [reply]
@ctalledo thanks for getting back to me - yes according to the docs Bottlerocket is based on kernel 5.4, so hopefully it will be compatible

Any idea of a likely timescale for Sysbox 0.5.0?

**Reply by ctalledo** [reply]
Hi @rcgeorge23, unfortunately 5.4 won't do (it does not carry the ID-mapped mounts feature); we need 5.12. 

Sysbox v0.5.0 release is due early March (we are working hard on it).

**Reply by StevenPJ** [reply]
Hi @ctalledo, bottlerocket have included [5.15 in their latest variants](https://github.com/bottlerocket-os/bottlerocket/blob/01d3a631f099b6c12098e18284d5c54cbcad0b71/variants/aws-k8s-1.24/Cargo.toml#L39). Have you guys been able to do any tests with it since its release?

**Reply by ctalledo** [reply]
Hi @StevenPJ, apologies for the belated reply (the message notification got buried in my email). 

We've not yet had a chance to try Sysbox on bottlerocket, but since it includes kernel 5.15, I don't expect too many issues to pop up. Have you had a chance to try it?

One thing that would help us support it would be to have a container image based on Botterocket, such that we can run Sysbox inside of it and catch any issues (similar to what we have for other Linux distros (see Dockerfiles [here](https://github.com/nestybox/sysbox/tree/master/tests)).

**Reply by stmcginnis** [reply]
> to have a container image based on Botterocket

One point of clarification - Bottlerocket is an OS for ~running~ containers, but as a base for container images themselves.

**Reply by ctalledo** [reply]
Hi @stmcginnis, thanks for the clarification. My point there was that some Linux distros offer a containerized image of the distro itself (just the user-space portion, not the kernel). For example, the Ubuntu or Alpine images. I am hoping there is something similar for Bottlerocket, as that helps us test that Sysbox works in the distro, with the caveat that the kernel in an actual Bottlerocket installation would be different.

**Reply by stmcginnis** [reply]
Unfortunately no. Bottlerocket is a container optimized OS (think CoreOS and Flatcar), so it's not something that can be containerized itself.

**Reply by ctalledo** [reply]
Oh that's too bad. FYI there is one such container image for Flatcar, we used it about a year ago to test Sysbox on it.

**Comment by eugenejen** [comment]
is there any roadmap on running sysbox on bottlerocket os now?

**Comment by ctalledo** [comment]
Hi @eugenejen, unfortunately we don't have concrete plans to support Sysbox on Bottlerocket, but I would be surprised if it does not work already. I say so because the latest version of Sysbox (v0.6.2) works on several Linux distros, assuming kernel >= 5.19. We don't have a package installer for distros other than Ubuntu/Debian, but it's pretty easy to build the binaries from source and then run them on any distro.

**Reply by eugenejen** [reply]
Hi @ctalledo,

Thanks for reply! 

I notice this issue in bottlerocket and I am not sure if affects sysbox https://github.com/bottlerocket-os/bottlerocket/issues/2163.

I recent looked for ways to run the sysbox daemonset distribution on bottlerocket os. but after i read the above issue, i wonder will that affect the goal we want to achieve (running a docker in sysbox under bottlerocket).

**Reply by ctalledo** [reply]
Hi @eugenejen, I read quickly through that bottlerocket issue and I don't believe it will apply to Sysbox (it automatically handles several of the issues described in there already). 

Question: are you planning to run K8s + Sysbox on bottlerocket? Or just Sysbox on bottlerocket? The former is a bit harder, I would start with the latter first.

**Reply by eugenejen** [reply]
@ctalledo  I am planning to run k8s + sysbox on top of bottlerocket. i do have a fallback now to run k8s + sysbox on ubuntu. I will let you know what i find.

**Reply by eugenejen** [reply]
Hi @ctalledo ,

I think the current amazon bottlerocket is configured to have /usr/local as read-only by default.  Therefore sysbox-install will failed as 

`  Warning  Failed     10s                kubelet            Error: failed to generate container "134d3472b419b5a66c495d7297c45f6b248a42105672a27f0053c5d4cbc60cda" spec: failed to apply OCI options: failed to mkdir "/usr/local/bin/": mkdir /usr/local: read-only file system`

**Reply by ctalledo** [reply]
Thanks @eugenejen; what would be the right directory to install binaries on Bottlerocket then? 

BTW, the sysbox-deploy-k8s daemonset source code is [here](https://github.com/nestybox/sysbox-pkgr/tree/master/k8s), and the main installation script is [here](https://github.com/nestybox/sysbox-pkgr/blob/master/k8s/scripts/sysbox-deploy-k8s.sh). We can adjust it for bottlerocket as needed.

**Reply by eugenejen** [reply]
@ctalledo I think i probably need to submit and MR. For the moment i just start to create an EKS with ubuntu.

I found a different issue now for ubutun but that will be different discussion topic.

**Reply by eugenejen** [reply]
just a note. the root cause for bottlerocket filesystem lockdown is
due to this feature.

https://github.com/bottlerocket-os/bottlerocket/blob/develop/SECURITY_FEATURES.md#immutable-rootfs-backed-by-dm-verity

**Reply by eugenejen** [reply]
@ctalledo from what i read from bottlerocket. it seems like unless we add sysbox into bottlerocket build pipeline, I don't see a good way. We may need to install all packages on data devices, but any mutation on systemd's configuration is probably impossible.
